pandas/_libs/reduction.pyx

```diff
@@ -499,7 +499,9 @@ def apply_frame_axis0(object frame, object f, object names,
                 # `piece` might not have an index, could be e.g. an int
                 pass
 
-            if not is_scalar(piece):
+            if isinstance(piece, list):
```

Comment:
this is not the correct patch, 

instead lets change line 507 (in revised code) to be
```
piece = deepcopy(piece)
```

Reply:
> this is not the correct patch,
> 
> instead lets change line 507 (in revised code) to be
> 
> ```
> piece = deepcopy(piece)
> ```

This causes some failures, e.g.
```
$ pytest pandas/tests/groupby/test_groupby.py::test_no_mutate_but_looks_like
```
which doesn't fail with
```
            if isinstance(piece, list):
                piece = deepcopy(piece)
            elif not is_scalar(piece):
                # Need to copy data to avoid appending references
                if hasattr(piece, "copy"):
                    piece = piece.copy(deep='all')
                else:
                    piece = copy(piece)
```

Reply:
@MarcoGorelli what is the difference that this causes in that test? Would it return one of the results completely as a list instead of exploding to a series?

Returning a list from a groupby op is thorny to begin with so I think should streamline logic as much as possible

Reply:
@WillAyd here's the difference:

```
>>> df = DataFrame({"key": [1, 1, 1, 2, 2, 2, 3, 3, 3], "value": range(9)})

>>> result1 = df.groupby("key", group_keys=True).apply(lambda x: x[:].key)
>>> result2 = df.groupby("key", group_keys=True).apply(lambda x: x.key)
>>> result1
key   
1    0    1
     1    1
     2    1
2    3    2
     4    2
     5    2
3    6    3
     7    3
     8    3

>>> result2
key  0  1  2
key         
1    1  1  1
2    2  2  2
3    3  3  3
```

> Returning a list from a groupby op is thorny to begin with so I think should streamline logic as much as possible

OK, I'll think of a workaround (but I realise that you'd like to get a v1.0.1 release out very soon so if you think this'll take me too long please do feel free to take over)

Reply:
I'm somewhat confused here. If I use this, the test case passes:
```
pd._testing.assert_series_equal(deepcopy(piece), piece.copy(deep="all"))
piece = piece.copy(deep="all")
```
but if I use this, it doesn't!
```
```
pd._testing.assert_series_equal(deepcopy(piece), piece.copy(deep="all"))
piece = deepcopy(piece)
```

will look into it

Reply:
@dsaxton see here - I share your puzzlement :) Will look into `Series.copy` tomorrow